homeassistant/components/vallox/__init__.py

```diff
@@ -0,0 +1,252 @@
+        SERVICE_SET_PROFILE_FAN_SPEED_HOME: {
+            'method': 'async_set_profile_fan_speed_home',
+            'schema': SERVICE_SCHEMA_SET_PROFILE_FAN_SPEED},
+        SERVICE_SET_PROFILE_FAN_SPEED_AWAY: {
+            'method': 'async_set_profile_fan_speed_away',
+            'schema': SERVICE_SCHEMA_SET_PROFILE_FAN_SPEED},
+        SERVICE_SET_PROFILE_FAN_SPEED_BOOST: {
+            'method': 'async_set_profile_fan_speed_boost',
+            'schema': SERVICE_SCHEMA_SET_PROFILE_FAN_SPEED},
+}
+
+DEFAULT_FAN_SPEED_HOME = 50
+DEFAULT_FAN_SPEED_AWAY = 25
+DEFAULT_FAN_SPEED_BOOST = 65
+
+
+async def async_setup(hass, config):
+    """Set up the client and boot the platforms."""
+    if DOMAIN not in config:
+        return True
+
+    conf = config[DOMAIN]
+    host = conf.get(CONF_HOST)
+    name = conf.get(CONF_NAME)
+
+    client = Vallox(host)
+    state_proxy = ValloxStateProxy(hass, client)
+    service_handler = ValloxServiceHandler(client, state_proxy)
+
+    hass.data[DOMAIN] = {
+        'client': client,
+        'state_proxy': state_proxy,
+        'name': name
+    }
+
+    for vallox_service in SERVICE_TO_METHOD:
+        schema = SERVICE_TO_METHOD[vallox_service]['schema']
+        hass.services.async_register(DOMAIN, vallox_service,
+                                     service_handler.async_handle,
+                                     schema=schema)
+
+    # Fetch initial state once before bringing up the platforms.
+    await state_proxy.async_update(None)
+
+    hass.async_create_task(async_load_platform(hass, 'sensor', DOMAIN, {}, config))
+    hass.async_create_task(async_load_platform(hass, 'fan', DOMAIN, {}, config))
+
+    return True
+
+
+class ValloxStateProxy:
+    """Helper class to reduce websocket API calls."""
+
+    def __init__(self, hass, client):
+        self._hass = hass
+        self._client = client
+        self._metric_cache = {}
+        self._profile = None
+        self._valid = False
+
+        async_track_time_interval(self._hass, self.async_update, SCAN_INTERVAL)
+
+    def fetch_metric(self, metric_key):
+        """Return cached state value."""
+        _LOGGER.debug("Fetching metric key: %s", metric_key)
+
+        if not self._valid:
+            raise IOError("Device state out of sync.")
+
+        if not metric_key in vlxDevConstants.__dict__.keys():
+            raise KeyError("Unknown metric key: {}".format(metric_key))
+
+        return self._metric_cache[metric_key]
+
+    def get_profile(self):
+        """Return cached profile value."""
+        _LOGGER.debug("Returning profile.")
+        if not self._valid:
+            raise IOError("Device state out of sync.")
+
+        return PROFILE_TO_STR_REPORTABLE[self._profile]
+
+    async def async_update(self, event_time):
+        """Fetch state update."""
+        _LOGGER.debug("Updating Vallox state cache.")
+
+        try:
+            self._metric_cache = self._client.fetch_metrics()
```

Comment:
If I'm reading the vallox library correctly, this does non asyncio I/O in a coroutine. That's not allowed. Coroutines will be executed in the event loop which should never block.

Reply:
Consider switching affected methods to regular sync functions. This might mean we have to use the home assistant sync api.

The alternative is to schedule I/O on the executor thread pool with `hass.async_add_executor_job`. But if we need to do that a lot, the sync api is cleaner.